runtime/gc_ctrl.c

```diff
@@ -103,20 +103,23 @@ CAMLprim value caml_gc_minor_words(value v)
 
 CAMLprim value caml_gc_counters(value v)
 {
-  CAMLparam0 ();   /* v is ignored */
-  CAMLlocal1 (res);
+  CAMLparam0 (); /* v is ignored */
+  CAMLlocal3 (minwords_, prowords_, majwords_);
 
   /* get a copy of these before allocating anything... */
   double minwords = caml_gc_minor_words_unboxed();
   double prowords = (double)Caml_state->stat_promoted_words;
   double majwords = Caml_state->stat_major_words +
                     (double) Caml_state->allocated_words;
 
-  res = caml_alloc_3(0,
-    caml_copy_double (minwords),
-    caml_copy_double (prowords),
-    caml_copy_double (majwords));
-  CAMLreturn (res);
+  minwords_ = caml_copy_double(minwords);
+  prowords_ = caml_copy_double(prowords);
+  majwords_ = caml_copy_double(majwords);
+  v = caml_alloc_small(3, 0);
```

Comment:
I would have used `value res = ...` instead of weirdly reusing the output variable, and then `caml_alloc_3` as before, but you've suffered enough roundtrips from me by now (today is the day I discovered the document listing [review antipatterns](https://www.chiark.greenend.org.uk/~sgtatham/quasiblog/code-review-antipatterns/) and I feel personally targeted by this list!) and I'm fine with the current version.

Reply:
Thanks!  I’d argue that `caml_alloc_N` functions are themselves antipatterns, as they must pointlessly reregister their arguments.  That’s why I went with `caml_alloc_small`.  Yes, users might well choose to trade performance for convenience, but I feel that the OCaml distribution itself should not.

Reply:
Can you actually measure a performance difference between the two options in a microbenchmark calling `Gc.counters`?

Reply:
Probably not.

Reply:
I nerd-sniped myself and did the measurements. Here is the benchmark:

```ocaml
let n = try int_of_string Sys.argv.(1) with _ -> failwith "expected an integer argument on the command line"

let () =
  for _ = 1 to n do
    ignore (Gc.counters ())
  done
```

And here are the versions compared:

```c
/* version: "trunk".
  This version is incorrect as the caml_copy_double() temporaries may get invalidated. */
CAMLprim value caml_gc_counters(value v)
{
  CAMLparam0 ();   /* v is ignored */
  CAMLlocal1 (res);

  /* get a copy of these before allocating anything... */
  double minwords = caml_gc_minor_words_unboxed();
  double prowords = (double)Caml_state->stat_promoted_words;
  double majwords = Caml_state->stat_major_words +
                    (double) Caml_state->allocated_words;

  res = caml_alloc_3(0,
    caml_copy_double (minwords),
    caml_copy_double (prowords),
    caml_copy_double (majwords));
  CAMLreturn (res);
}

/* version "demi" */
CAMLprim value caml_gc_counters(value v)
{
  CAMLparam0 ();   /* v is ignored */
  CAMLlocal3 (minwords_, prowords_, majwords_);

  /* get a copy of these before allocating anything... */
  double minwords = caml_gc_minor_words_unboxed();
  double prowords = (double)Caml_state->stat_promoted_words;
  double majwords = Caml_state->stat_major_words +
                    (double) Caml_state->allocated_words;

  minwords_ = caml_copy_double(minwords);
  prowords_ = caml_copy_double(prowords);
  majwords_ = caml_copy_double(majwords);
  value res = caml_alloc_small(3, 0);
  Field (res, 0) = minwords_;
  Field (res, 1) = prowords_;
  Field (res, 2) = majwords_;
  CAMLreturn (res);
}

/* version "store_field": following the style around in the file */
CAMLprim value caml_gc_counters(value v)
{
  CAMLparam0 ();   /* v is ignored */
  CAMLlocal3 (minwords_, prowords_, majwords_);

  /* get a copy of these before allocating anything... */
  double minwords = caml_gc_minor_words_unboxed();
  double prowords = (double)Caml_state->stat_promoted_words;
  double majwords = Caml_state->stat_major_words +
                    (double) Caml_state->allocated_words;

  minwords_ = caml_copy_double(minwords);
  prowords_ = caml_copy_double(prowords);
  majwords_ = caml_copy_double(majwords);
  value res = caml_alloc_small(3, 0);
  Store_field (res, 0, minwords_);
  Store_field (res, 1, prowords_);
  Store_field (res, 2, majwords_);
  CAMLreturn (res);
}


/* version "clean": the simplest possible */
CAMLprim value caml_gc_counters(value v)
{
  CAMLparam0 ();   /* v is ignored */
  CAMLlocal4 (minwords_, prowords_, majwords_, res);

  /* get a copy of these before allocating anything... */
  double minwords = caml_gc_minor_words_unboxed();
  double prowords = (double)Caml_state->stat_promoted_words;
  double majwords = Caml_state->stat_major_words +
                    (double) Caml_state->allocated_words;

  minwords_ = caml_copy_double(minwords);
  prowords_ = caml_copy_double(prowords);
  majwords_ = caml_copy_double(majwords);
  res = caml_alloc_3(0, minwords_, prowords_, majwords_);
  CAMLreturn (res);
}
```

And the benchmark results:

| Command       |    Mean [ms] | Min [ms] |    Relative |
|:--------------|-------------:|---------:|------------:|
| `trunk`       |  243.3 ± 2.9 |    238.7 |        1.00 |
| `clean`       | 274.1 ± 19.2 |    262.7 | 1.13 ± 0.08 |
| `demi`        | 286.7 ± 26.3 |    267.8 | 1.18 ± 0.11 |
| `store_field` | 359.3 ± 25.2 |    341.6 | 1.48 ± 0.11 |

It turns out that the `clean` version I was suggesting is in fact slightly faster than your version (in ways that probably don't matter in practice).

Reply:
Good catch!  I’m curious why that is, but it’s definitely the way to go.

Reply:
I also looked and I'm not sure. The code of `caml_alloc_3` is careful to only register more roots *if* a collection is needed, so in fact there is no additional rooting overhead. I played a bit with godbolt to understand why the version was slightly faster instead of same-speed, and I noticed differences in spilling (maybe spilling double temporaries is more costly), but nothing that clearly explains the small gap.